feat(trends): add moving average `smoothing_intervals` param

hazzadous opened the pull request:
This will average over the preceding `smoothing_intervals` intervals.
The driver for this is to remove either weekly or monthly trends.

Limitations at the moment are that it only works for total volume
queries, as it relies on the results from clickhouse being from a single
timeline series.


Video example [here](https://www.loom.com/share/56eb5f031b6e448288df9f2fb5a1bac6)

Commit:
feat(trends): add moving average `smoothing_intervals` param

This will average over the preceding `smoothing_intervals` intervals.
The driver for this is to remove either weekly or monthly trends.

Limitations at the moment are that it only works for total volume
queries, as it relies on the results from clickhouse being from a single
timeline series.

MarconLP committed:
WIP UI for selecting smoothing

MarconLP committed:
Merge branch 'feat/trend-smoothing' of https://github.com/PostHog/posthog into feat/trend-smoothing

MarconLP committed:
WIP: Fix typing and appearce logic

MarconLP committed:
Made smoothing extensible beyond day

marcushyett-ph committed:
Test now considers a non-trivial case

Commit:
chore(smoothing): use url params for state storage of smoothing value

This should allow for url navigation to work, although I haven't quite
figured out how we then use this to make the insights/trends POST
request

marcushyett-ph committed:
Added further tests for smoothing intervals

Commit:
fix test_to_dict

Commit:
fix trend tests

marcushyett-ph committed:
Added test to validate preceeding limit

marcushyett-ph committed:
Fixed test for 2 interval

Commit:
clear django redis cache before each test

Commit:
Merge remote-tracking branch 'origin/master' into feat/trend-smoothing

Commit:
add explicit ordering to dashboard items in refresh test

Otherwise the test is not deterministic.

Commit:
Revert "clear django redis cache before each test"

This is a good thing to do, but I don't want to open up a discussion on
this PR. It was not fix for the immediate intermittent
test_refresh_cache test

This reverts commit aa52f332f35c97076dcd1ec5e608e1450e83832d.

Commit:
format

Commit:
clickhouse trends: SET allow_experimental_window_functions = 1

Commit:
fix setting of allow_experimental_window_functions

neilkakkar committed:
add cache clears for requests

neilkakkar reviewed (approved):
Few small comments, but LGTM!

marcushyett-ph committed:
Delete SmoothingFilterLogic.ts

Removed smoothing filter logic file

Commit:
Use `SMOOTHING_INTERVALS` instead of string literal

This should ensure that if we update the url param key somewhere, we don't have to remember to change it here as well.

Co-authored-by: Neil Kakkar <neilkakkar@gmail.com>

Commit:
Merge remote-tracking branch 'origin/master' into feat/trend-smoothing

Commit:
move test_trends to ee

Commit:
actually delete test_trends.py

mariusandra reviewed (changes requested):
Apologies for the wait. 

Went through the code, and my general suggestion is that whenever building something new, always look at the code around you to see how other similar things have been done. For example look at `IntervalFilter`. Then follow that pattern. That leads to the least surprises and most predictable behaviour, and faster code reviews.

Here you're again introducing a completely new way of doing things, by creating hook-adapters that mix logics with components, and create selectors that are not selectors and actions that are not actions. No, stop that :D. Let's unify the data layer.

However, there's an issue. All the existing frontend filters follow a bad pattern, where they are [connected to the URL](https://github.com/PostHog/posthog/blob/00633412cc9e2ead59a6a538468d452e06b0c9ef/frontend/src/lib/components/IntervalFilter/intervalFilterLogic.ts#L49). That's causing [issues with turbo mode](https://github.com/PostHog/posthog/issues/6513#issuecomment-952279054) for example. The right solution to either connect each filter directly to `insightsLogic`, or to a custom logic (`smoothingFilterLogic`) that provides custom actions (`setSmoothingFilter`) that in turn call setFilters with the right params. Kind of how [trendsLogic](https://github.com/PostHog/posthog/blob/eda638a3b49cd84bbe53a7e1d899c542a82bb792/frontend/src/scenes/trends/trendsLogic.ts#L14) and funnelLogic do it.

marcushyett-ph commented:
Thanks @mariusandra. @hazzadous let me know if you fancy pairing on this tomorrow or Friday?

hazzadous commented:
Yup tomorrow works for me

marcushyett-ph commented:
I'll find some time.

MarconLP committed:
Transitioned smoothing logic to use Logic

marcushyett-ph commented:
@mariusandra thanks for the feedback - we transitioned this to use logic - referencing IntervalFilterLogic.

We've come across one issue, we'd love your help with though. We've connected SmoothingFilterLogic to IntervalFilterLogic, so the smoothing is updated to match the interval being used.

However it seems to create some kind of infinite loop on SetIntervalFilter, would appreciate any pointers.

cc: @hazzadous to share any more context

mariusandra commented:
I wrote this the last time around:

> However, there's an issue. All the existing frontend filters follow a bad pattern, where they are [connected to the URL](https://github.com/PostHog/posthog/blob/00633412cc9e2ead59a6a538468d452e06b0c9ef/frontend/src/lib/components/IntervalFilter/intervalFilterLogic.ts#L49). That's causing [issues with turbo mode](https://github.com/PostHog/posthog/issues/6513#issuecomment-952279054) for example. The right solution to either connect each filter directly to `insightsLogic`, or to a custom logic (`smoothingFilterLogic`) that provides custom actions (`setSmoothingFilter`) that in turn call setFilters with the right params. Kind of how [trendsLogic](https://github.com/PostHog/posthog/blob/eda638a3b49cd84bbe53a7e1d899c542a82bb792/frontend/src/scenes/trends/trendsLogic.ts#L14) and funnelLogic do it.

I suggest you remove the `urlToAction` / `actionToUrl` code from inside this new logic, and instead follow the pattern trendsLogic/funnelsLogic use: connect to `setFilters` and `filters` from `insightLogic`. Then use selectors to get the current value of the filter, and have your own `setSmoothing` action that calls `setFilters` with the right params.

MarconLP committed:
WIP: Moving away from url based logic

MarconLP committed:
WIP: Refactored logic on listeners

Commit:
fix storybook

Commit:
refactor

marcushyett-ph commented:
TODO: When return from hols - refactor logic for interval to match new logic for smoothing so they are in sync.

marcushyett-ph commented on frontend/src/lib/components/SmoothingFilter/smoothingFilterLogic.ts:
```diff
@@ -16,7 +17,9 @@ export const smoothingFilterLogic = kea<smoothingFilterLogicType>({
     listeners: ({ actions, props }) => ({
         setSmoothing: async ({ filter }) => {
-            insightLogic(props).actions.setFilters({ smoothing_intervals: filter })
+            insightLogic(props).actions.setFilters({
```
This should actually call `actions` which is connected above rather than another instance of insightLogic
```suggestion
            actions.setFilters({
```

marcushyett-ph replied:
But it wont work until we refactor the IntervalLogic

Commit:
Merge remote-tracking branch 'origin/master' into feat/trend-smoothing

This merge is pretty big and I wasn't totally confident that it was
completely correct. I did a pretty shoddy job of merging the
test_clickhouse_trends.py file together.

Commit:
set redis url

Commit:
remove keastory

marcushyett-ph committed:
Move interval filter to setFilters insight logic

marcushyett-ph commented:
@hazzadous fancy pairing on some text / merge conflict fixing?

hazzadous commented:
> @hazzadous fancy pairing on some text / merge conflict fixing?

yup let me know when you're available

marcushyett-ph committed:
Merge remote-tracking branch 'origin/master' into feat/trend-smoothing

marcushyett-ph commented:
@hazzadous lets try and get this out tomorrow in the wework?

hazzadous commented:
100 %

Commit:
Merge remote-tracking branch 'origin/master' into feat/trend-smoothing

Commit:
update trend snapshots

Commit:
run prettier

Commit:
Add stickiness test deps that were in test_trends

Commit:
Fix storybook

Commit:
fix stickiness tests

Commit:
update snapshot

Commit:
fix cohort test

Commit:
only display smoothing on linear

Commit:
fix tests

Commit:
Merge remote-tracking branch 'origin/master' into feat/trend-smoothing

Commit:
Add some tests

mariusandra commented on frontend/src/lib/components/IntervalFilter/IntervalFilter.tsx:
```diff
@@ -6,15 +6,19 @@ import { disableHourFor, disableMinuteFor } from 'lib/utils'
+    const {
+        filters: { interval },
+    } = useValues(intervalFilterLogic(insightProps))
+    const { setIntervalFilter, setDateFrom } = useActions(intervalFilterLogic(insightProps))
```
This `setDateFrom` got recently removed in https://github.com/PostHog/posthog/pull/7758 ... when all that code got moved into a listener.

mariusandra commented on frontend/src/lib/components/SmoothingFilter/smoothingFilterLogic.ts:
```diff
@@ -0,0 +1,35 @@
+            if (!interval) {
+                return
+            }
+            if (!smoothingOptions[interval].find((option) => option.value === smoothing_intervals)) {
+                if (smoothing_intervals !== 1) {
+                    actions.setSmoothing(1)
+                }
+            }
```
Could this instead go in `cleanFilters`?

mariusandra commented on frontend/src/scenes/insights/InsightTabs/InsightDisplayConfig.tsx:
```diff
@@ -116,6 +123,13 @@ export function InsightDisplayConfig({ filters, activeView, disableTable }: Insi
+                {activeView === InsightType.TRENDS &&
+                !filters.breakdown_type &&
+                !filters.compare &&
+                (!filters.display || filters.display === ACTIONS_LINE_GRAPH_LINEAR) ? (
```
Could we also add this to the cumulative chart... or not really worth it?

hazzadous replied:
We decided go narrow on this PR, to reduce chances of further iterations.

hazzadous replied:
And also, not as relevant

mariusandra commented on posthog/api/test/test_trends.py:
```diff
@@ -1,209 +0,0 @@
-import json
```
Where did this go?

hazzadous replied:
I think it was just an unused import 🤞

mariusandra replied:
I mean the entire file :D

hazzadous replied:
hah yup it's moved to under `ee/`

Commit:
Merge remote-tracking branch 'origin/master' into feat/trend-smoothing

mariusandra reviewed (dismissed):
Overall, looks really good! 

I found a few comments... and this should be merged with `master`, as one of the improvements here already got built in with https://github.com/PostHog/posthog/pull/7758 (yay for WIP)

Finally, 
1) can we make this dropdown also a "box" like the other dropdowns?
2) it's a bit strange to see "123.456" pageviews. Could we add a "* averaged over 7 days" label to the tooltip somehow? 

![2021-12-17 14 30 21](https://user-images.githubusercontent.com/53387/146551977-9cbb9bc6-82c8-4ead-9b2c-2e7510882e6b.gif)

mariusandra commented:
@hazzadous Just a few things left in this PR that's been open from October, shame to close it now.

marcushyett-ph commented:
@mariusandra this one is on me to get finished... have time in my diary Thursday morning to get these last bits sorted :)

hazzadous commented:
❤️

marcushyett-ph committed:
added box to make dropdown consistent

marcushyett-ph committed:
Co-authored-by: Harry Waye <harry@microwayes.net>

marcushyett-ph commented:
Thanks @mariusandra! Every time you re-open this I get more embarrassed about my incredibly slow pace on this PR. Will try harder.

hazzadous commented:
I'm going to close this, if someone picks the task up they can open up a PR themselves.

EDsCODE commented:
Cool. Going to pick this up within the next week or so. Leaving a note here for record

timgl committed:
Merge branch 'master' into feat/trend-smoothing

timgl committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/trend-smoothing

timgl committed:
Floor

timgl committed:
fix tests

timgl committed:
fix tests

timgl committed:
fix test

EDsCODE commented on frontend/src/scenes/insights/InsightTabs/InsightDisplayConfig.tsx:
```diff
@@ -117,6 +118,13 @@ export function InsightDisplayConfig({
+                !filters.breakdown_type &&
+                !filters.compare &&
+                (!filters.display || filters.display === ACTIONS_LINE_GRAPH_LINEAR) ? (
+                    <SmoothingFilter />
```
Shall we feature flag? At least for failure control incase something is broken?

mariusandra commented on frontend/src/lib/components/SmoothingFilter/smoothings.ts:
```diff
@@ -0,0 +1,35 @@
+            value: 28,
+        },
+    ],
+    week: [],
```
As someone who has extensively used smoothing over my last 8 years in my last startup ([internal issue](https://github.com/PostHog/product-internal/issues/68)), these day values are really the only values I ever used.

timgl committed:
Add feature flag

timgl committed:
Merge branch 'master' into feat/trend-smoothing

timgl commented on frontend/src/scenes/insights/InsightTabs/InsightDisplayConfig.tsx:
```diff
@@ -117,6 +118,13 @@ export function InsightDisplayConfig({
+                !filters.breakdown_type &&
+                !filters.compare &&
+                (!filters.display || filters.display === ACTIONS_LINE_GRAPH_LINEAR) ? (
+                    <SmoothingFilter />
```
Done
